eventuals/type-traits.h

```diff
@@ -298,6 +303,104 @@ using apply_tuple_types_t = typename apply_tuple_types<T, Tuple>::type;
 
+template <typename... Errors>
+inline constexpr bool check_errors_v =
+    std::conjunction_v<
+        std::is_base_of<std::exception, std::decay_t<Errors>>...>;
+
+template <typename... Errors>
+using check_errors_t =
+    std::conditional_t<
+        check_errors_v<Errors...>,
+        std::true_type,
+        std::false_type>;
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename T>
+struct is_variant : std::false_type {};
+
+template <typename... Args>
+struct is_variant<std::variant<Args...>> : std::true_type {};
+
+template <typename T>
+inline constexpr bool is_variant_v = is_variant<T>::value;
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename T>
+struct is_tuple : std::false_type {};
+
+template <typename... Args>
+struct is_tuple<std::tuple<Args...>> : std::true_type {};
+
+template <typename T>
+inline constexpr bool is_tuple_v = is_tuple<T>::value;
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename V, typename T>
+struct tuple_contains_exact_type;
+
+template <typename V, typename T0, typename... T>
+struct tuple_contains_exact_type<V, std::tuple<T0, T...>> {
+  static constexpr bool value = tuple_contains_exact_type<V, std::tuple<T...>>::value;
+};
+
+template <typename V, typename... T>
+struct tuple_contains_exact_type<V, std::tuple<V, T...>> {
+  static constexpr bool value = true;
+};
+
+template <typename V>
+struct tuple_contains_exact_type<V, std::tuple<>> {
+  static constexpr bool value = false;
+};
+
+template <typename V, typename T>
+inline constexpr bool tuple_contains_exact_type_v = tuple_contains_exact_type<V, T>::value;
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename>
+struct check_variant_errors;
+
+template <typename... Errors>
+struct check_variant_errors<std::variant<Errors...>> {
+  static constexpr bool value = check_errors_v<Errors...>;
+};
+
+template <typename Variant>
+inline constexpr bool check_variant_errors_v = check_variant_errors<Variant>::value;
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename T>
+struct LambdaType : public LambdaType<decltype(&T::operator())> {};
+
+template <typename ClassType, typename ReturnType, typename... Args>
+struct LambdaType<ReturnType (ClassType::*)(Args...) const> {
+  static constexpr int size = sizeof...(Args);
+
+  template <size_t i>
+  struct arg {
+    typedef typename std::tuple_element<i, std::tuple<Args...>>::type type;
+  };
+
+  typedef ReturnType result_type;
+};
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename T>
+using has_operator = decltype(&T::operator());
+
+template <typename T>
+using is_default_lambda =
+    std::experimental::is_detected<has_operator, T>;
+
+////////////////////////////////////////////////////////////////////////
+
```

Comment:
Just flagging that this is likely dead code now given our discussion in `catch.h`.